### PR TITLE
fix(): Async GIT due to performance issues

### PR DESCRIPTION
**Pull request by marhi**
The GIT service was partially rewritten for better performance mostly on MacOS.
There is still some work to be done, like changing the fileExistsSync to async version.

### CODE REVIEW

File: src/app/components/window-bottom/window-bottom.component.ts [hunk 1]
```diff
@@ -55,8 +55,8 @@ export class WindowBottomComponent implements OnInit {
           this.currentProcess = this.xterm.terminals[event.index].title;
           this.currentDir = this.xterm.terminals[event.index].dir;
           this._git.dir = this.currentDir;
-          this.currentBranch = this._git.branch;
-          this.currentStatus = this._git.status;
+          this._git.branch.then(res => this.currentBranch = res);
```
Comment (jkuri) [inline src/app/components/window-bottom/window-bottom.component.ts]: 👍

File: src/app/services/git.service.ts [hunk 2]
```diff
@@ -4,57 +4,47 @@ import * as os from 'os';
...
-      return true;
-    } catch (e) {
-      this.currentStatus = null;
-      this.currentBranch = null;
-      return false;
-    }
-  }
-
   parseStatus(s: string): TGitStatus {
+    if (s === null) { return '?'; }
```
Comment (jkuri) [inline src/app/services/git.service.ts]:
```ts
if (!s) { return '?'; }
```

File: src/app/services/git.service.ts [hunk 3]
```diff
@@ -4,57 +4,47 @@ import * as os from 'os';
...
   }
 
-  get status(): TGitStatus { return this.currentStatus; }
-  set status(s: TGitStatus) { this.currentStatus = s; }
-  set branch(b: string) { this.currentBranch = b; }
-  get branch(): string { return this.currentBranch; }
+  get status(): Promise<TGitStatus> {
+    if (!this.cwd) { return Promise.resolve('?'); }
+    return this.git.from(this.cwd).getStatus().then(status => {
+        return Promise.resolve(!status.code ? this.parseStatus(status.value) : '?');
```
Comment (jkuri) [inline src/app/services/git.service.ts]:
```ts
return this.git.from(this.cwd).getStatus().then(status => !status.code ? this.parseStatus(status.value) : '?');
```

if you have luck you could squeze this into oneline :)

File: src/app/services/git.service.ts [hunk 4]
```diff
@@ -4,57 +4,47 @@ import * as os from 'os';
...
+    });
+  }
+
+  get branch(): Promise<string> {
+    if (!this.cwd) { return Promise.resolve(null); }
+    return this.git.from(this.cwd).getBranch().then(branch => {
+        return Promise.resolve(!branch.code ? branch.value : null);
```
Comment (jkuri) [inline src/app/services/git.service.ts]: check previous comment.

File: src/app/services/git.service.ts [hunk 5]
```diff
@@ -79,42 +68,44 @@ class GITHandler {
...
+  gitExec(params: string[]): Promise<IGitResult> {
+    return new Promise<IGitResult>(resolve => {
+      if (!this.gitBin || !fs.existsSync(this.dir)) { resolve({ code: null, value: null }); }
+      let log: string = '';
+      let options: child_process.SpawnSyncOptionsWithStringEncoding = {
+        timeout: 100,
+        encoding: 'utf8',
+        env: process.env,
+        cwd: this.dir,
+        stdio: 'pipe'
```
Comment (jkuri) [inline src/app/services/git.service.ts]: why `stdio: pipe`?

File: src/app/services/git.service.ts (hunk 2 above)
Comment (marhi) [inline src/app/services/git.service.ts]: The reason for this:

> let s = ""
undefined
> s
''
>
> let a = null
undefined
> a
null
>
> !s
true
> !a
true
> s === a
false

File: src/app/services/git.service.ts (hunk 5 above)
Comment (marhi) [inline src/app/services/git.service.ts]: Sorry, I forgot that's the default.

File: src/app/services/git.service.ts (hunk 3 above)
Comment (marhi) [inline src/app/services/git.service.ts]: I had luck :)

**Pull request merged**